backend/src/main/java/gov/cdc/usds/simplereport/api/model/filerow/TestResultRow.java

```diff
@@ -447,11 +452,33 @@ public TestResultRow(Map<String, String> rawRow) {
   private List<FeedbackMessage> validateDeviceModelAndTestPerformedCode(
       String equipmentModelName, String testPerformedCode) {
 
-    if (validModelTestPerformedCombination(equipmentModelName, testPerformedCode)
-        || validDiseaseTestPerformedLoinc(testPerformedCode)) {
-      return emptyList();
+    if (equipmentModelName == null || testPerformedCode == null) {
+      return generateInvalidDataErrorMessages();
     }
 
+    ImmutablePair<Boolean, Boolean> deviceInDbPair =
+        validDeviceInDb(equipmentModelName, testPerformedCode);
+    boolean genericDeviceInDb = deviceInDbPair.getLeft();
+    boolean hivDeviceInDbAndActive = deviceInDbPair.getRight();
+
+    ImmutablePair<Boolean, Boolean> deviceInAllowListPair =
+        validDeviceInAllowList(testPerformedCode);
+    boolean genericDeviceInAllowList = deviceInAllowListPair.getLeft();
+    boolean hivDeviceInAllowListAndActive = deviceInAllowListPair.getRight();
+
+    boolean shouldReturnInvalidComboError = !(genericDeviceInDb || genericDeviceInAllowList);
+    boolean shouldReturnInactiveDiseaseError =
+        !(hivDeviceInDbAndActive || hivDeviceInAllowListAndActive);
+
+    if (shouldReturnInvalidComboError) {
+      return generateInvalidDataErrorMessages();
+    } else if (shouldReturnInactiveDiseaseError) {
+      return generateInactiveDiseaseErrorMessages();
```

Comment:
I think there's an opportunity for us to simplify this method to improve readability and avoid coupling it to HIV specifically. Right now the method has two spots where we return the invalid data error message, but the only boolean variables which affect that are the generic device ones so it seems that the invalid data error message can be handled separately from the inactive disease error checking.

For the example below, it will return the invalid data message if it is neither in the db nor in the allow list. If we handle the inactive disease message after that, then we know that the device data is technically valid because it is either in the db or in the allow list. At this point the only thing that would make it invalid is if it has an inactive disease so we can handle that case then.

```
        if (!validDeviceInDb(equipmentModelName, testPerformedCode)
                && !validDeviceInAllowList(testPerformedCode)) {
            return generateInvalidDataErrorMessages();
        }

        boolean hasOnlyActiveDiseases = resultsUploaderDeviceService.validateResultsOnlyIncludeActiveDiseases(equipmentModelName, testPerformedCode);
        if (!hasOnlyActiveDiseases) {
            return generateInactiveDiseaseErrorMessages();
        }

        return emptyList();
```

Note that in this case `validDeviceInDb` and `validDeviceInAllowList` would simply return a boolean which I think also improves readability and maintains a single responsibility for those methods rather than concerning them with which diseases are inactive. If we end up having to add similar functionality for syphilis in the future, we can simply update `validateResultsOnlyIncludeActiveDiseases` to check for whether syphilis is active or not for bulk upload. Let me know what your thoughts on this!

Reply:
this is much cleaner! will do: thanks so much